### PR TITLE
[MRG] Improve interpolation of bridged electrodes

### LINKED ISSUE #11079: Interpolate bridge electrodes when more than 2 electrodes are bridged together
Following @alexrockhill addition of https://mne.tools/stable/auto_examples/preprocessing/eeg_bridging.html and #10587

What about the case where more than 2 electrodes are bridged together, e.g. where a trio of electrodes are bridged together?

https://github.com/mne-tools/mne-python/blob/5ee892f2ef79f451b7bdba251f4dec18e266001e/mne/preprocessing/interpolate.py#L75-L155 

`interpolate_bridged_electrodes` only works on bridge pair by pair. But if we take the example below, we can do better for the bridge ('P5', 'PO7', 'PO3', 'O1'). 

image

I propose 2 changes to `interpolate_bridged_electrodes`: 
- handles a group of electrodes bridged together as a single bridge and computes a single virtual channel per group at the average location.
- add a warning if you use `interpolate_bridged_electrodes` where you have bridges including more than N electrodes. 

I think a good number would be N=3, a trio of electrodes is usually either a straight line on a row/column of the cap, or a triangle. If it's a row/column (e.g. PO8, PO6, PO4) the average location is actually the center electrode (PO6). If it's a triangle, the average location is more or less equidistant from the 3 electrodes. In both cases, intuitively, I think the virtual channel will improve interpolation. Above 3 electrodes, I'm not convinced. 

-------------------

To retrieve the group of electrodes, I can think of an implementation as a graph with `networkx`, something along the lines of:

```
G = nx.Graph()
for bridge in bridged_idx:
    G.add_edge(*bridge)
groups_idx = nx.connected_components(G)
```

Any other idea to avoid the dependency on `networkx`?

### LINKED ISSUE #11088: Bug with transformation from cartesian to sphere to cartesian coordinates
I am not sure it's a bug, I feel like I'm missing something obvious.. Following #11079 I've been looking at the code of `interpolate_bridged_electrodes` and I ran into a weird case. I'm testing code to add virtual channels at the average location between N bridged electrodes on this recording:

Screenshot 2022-08-25 at 11 48 24

I'm not going to provide that code snippet for now, I'll wait for opinions on #11079, but it creates 3 virtual channels. One at the center of the triangle (AF3, F3, F1), one very close to PO5... and a last one, between T8 and TP8 on the wrong side of the head!!

-------------

Now, consider this example, with code very similar to `interpolate_bridged_electrodes`:

```
import numpy as np
from mne import create_info
from mne.transforms import _cart_to_sph, _sph_to_cart


info = create_info(["T7", "TP7", "T8", "TP8"], sfreq=1024, ch_types="eeg")
info.set_montage("standard_1020")
montage = info.get_montage()
pos = montage.get_positions()
assert pos["coord_frame"] == "head"

ch_pos = pos["ch_pos"]

# create virtual channel midway between T7 and TP7
ch_names = ["T7", "TP7"]
sphere_positions = np.zeros((len(ch_names), 3))
for i, ch_name in enumerate(ch_names):
    sphere_positions[i, :] = _cart_to_sph(ch_pos[ch_name])
pos_virtual = _sph_to_cart(np.average(sphere_positions, axis=0))
```

`ch_pos` reads as:

```
OrderedDict([('T7',  array([-0.08598208,  0.01487164,  0.03117337])),
             ('TP7', array([-0.08675728, -0.01495098,  0.03515866])),
             ('T8',  array([0.08326137, 0.01525819, 0.03097297])),
             ('TP8', array([ 0.08362232, -0.01508567,  0.0350577 ]))])
```

Note that T7 and TP7, with odd numbers, have a **negative** x coordinate.

`pos_virtual` reads as `array([[ 8.76692427e-02, -2.68577709e-05,  3.31532296e-02]])`, with a **positive** x coordinate.

-----------

And to make it more explicit, here is the next couples of lines to get to a plot, using very similar code to `interpolate_bridged_electrodes`:

```
import numpy as np
from mne import create_info
from mne.io import RawArray
from mne.transforms import _cart_to_sph, _sph_to_cart


info = create_info(["T7", "TP7", "T8", "TP8"], sfreq=1024, ch_types="eeg")
info.set_montage("standard_1020")
montage = info.get_montage()
pos = montage.get_positions()
assert pos['coord_frame'] == "head"

ch_pos = pos["ch_pos"]

# create virtual channel midway between T7 and TP7
ch_names = ["T7", "TP7"]
sphere_positions = np.zeros((len(ch_names), 3))
for i, ch_name in enumerate(ch_names):
    sphere_positions[i, :] = _cart_to_sph(ch_pos[ch_name])
pos_virtual = _sph_to_cart(np.average(sphere_positions, axis=0))

# create virtual channel info
virtual_info = create_info(
    ch_names=["vT7-TP7"],
    sfreq=info["sfreq"],
    ch_types="eeg",
)
virtual_info["chs"][0]["loc"][:3] = pos_virtual

# create a RawArray to concatenate both info
raw = RawArray(np.random.randn(4, 1024), info)
raw_virtual = RawArray(np.random.randn(1, 1024), virtual_info)
raw.add_channels([raw_virtual], force_update_info=True)

# plot montage
raw.get_montage().plot()
```

Screenshot 2022-08-25 at 12 00 53

Interestingly, changing `ch_names` to `["T8", "TP8"]` does create the correct virtual channel between T8 and TP8.

### PR DESCRIPTION
**Pull request by mscheltienne**
Closes #11079 
Closes #11088

On main, `interpolate_bridged_electrodes` creates one virtual channel for each **pair** of bridged electrodes. On this PR, it creates one virtual channel for each **group** of bridged electrodes. A warning is emitted if the group includes more than 4 electrodes (not included).

On main, the virtual channel location is determined by averaging the 3 spherical coordinates of each pair of electrodes. On this PR, the location is determined by averaging the 3 cartesian coordinates of each electrode in the group. The resulting position is converted to spherical coordinates and the radius is replaced with the average radius of the electrodes in the group before being converted back to cartesian coordinates.

I ran this example locally, just in case: https://mne.tools/stable/auto_examples/preprocessing/eeg_bridging.html#ex-eeg-bridging

### CODE REVIEW

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [inline mne/preprocessing/interpolate.py:134]: This snippet looks for group of electrodes. It's equivalent to the `networkx` code:

```
G = nx.Graph()
for bridge in bridged_idx:
    G.add_edge(*bridge)
groups_idx = nx.connected_components(G)
```

I had no idea `scipy` had those functionalities, and I never used them. Maybe there is a nicer way to do this?
Also, I used `connected_components(csr_matrix(G_dense), directed=False)` as per the example of `connected_components`, but it also works without `csr_matrix`. Not sure if it's needed..

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [inline mne/preprocessing/interpolate.py:141]: I chose to emit the warning from N=5 onwards. Maybe it's a bit too lenient. LMK. 
Also, maybe we could improve a bit the warning, what would be a safe course of action we could recommend when you have a large bridged area? I would drop the electrodes, or at least most of the electrodes.

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [inline mne/preprocessing/interpolate.py:226]: I defined this in a separate function because I wanted to add a test specifically for this. Maybe this function and the associated test could be moved to a different module? Not sure where would be a good fit..

File: mne/preprocessing/interpolate.py
Comment (larsoner) [inline mne/preprocessing/interpolate.py:128]: I would do this with https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html

Or if you know the matrices will always be small (they probably will) then what you have is fine. I wouldn't bother casting explicitly to `csr_matrix`, though

File: mne/preprocessing/interpolate.py
Comment (larsoner) [reply mne/preprocessing/interpolate.py:140]: You could add a `bad_limit=4` param. It makes it explicit to people that there is a limit they should think about. And I'd maybe an error if it's exceeded, since they can work around it by setting it higher themselves

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [reply mne/preprocessing/interpolate.py:127]: It will always be small, a worst (impossible) case scenario would be a 256 electrodes cap with all electrodes bridged. So 256x256.

So `coo_matrix` to avoid the for loop and remove the `csr_matrix`. I'll check it later today.

File: mne/preprocessing/interpolate.py
Comment (larsoner) [inline mne/preprocessing/interpolate.py:217]:
```suggestion
    cartesian_positions = np.array([
        ch_pos[ch_name] for ch_name in group_names])
    sphere_positions = _cart_to_sph(cartesian_positions)
```

File: mne/preprocessing/interpolate.py
Comment (larsoner) [reply mne/preprocessing/interpolate.py:225]: I think this is an okay place for it, we can move if it's useful elsewhere

File: mne/preprocessing/interpolate.py
Comment (larsoner) [reply mne/preprocessing/interpolate.py:127]: Up to you if you want to `coo_matrix` or not. It looks like you might have to loop to get `idx0` and `idx1`. In principle you could maybe use `np.searchsorted` to avoid even those loops but it starts to be less and less readable...

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [reply mne/preprocessing/interpolate.py:140]: Sounds good, I'll add it.

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [reply mne/preprocessing/interpolate.py]: I don't mind, but we will definitely forget it 😄

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [reply mne/preprocessing/interpolate.py]: I'll stick with the current version, I just changed the `.index` with a better `np.searchsorted`. I did not find an easy and readable way to convert the information to the format for `coo_matrix`. I was starting with:

```
nodes = sorted(set(chain(*bridged_idx)))
rows, cols = zip(*bridged_idx)
# need to convert the idx as with `nodes.index(idx)`
coo_matrix(np.ones(len(rows)), rows, cols, shape=(len(nodes), len(nodes)))
# ...
```

It doesn't look more readable, and the conversion of the `idx` requires the same loop.

**Comment by mscheltienne** [comment]
Any way to kill `linux / pip-pre / linux pip 3.10`? It's been running for 5 hours..
How about adding `timeout-minutes: 120` here https://github.com/mne-tools/mne-python/blob/bdc435d8a7643a934365a221ab9d52ec8b33bc33/.github/workflows/linux_pip.yml#L18-L20

File: mne/preprocessing/interpolate.py
Comment (mscheltienne) [inline mne/preprocessing/interpolate.py:94]:
```suggestion
        (included) and interpolated. Above this number, an error will be
```

File: mne/preprocessing/interpolate.py
Comment (larsoner) [inline mne/preprocessing/interpolate.py:76]: We don't have any other type annotations, so let's not start here. Let's discuss separately if we should do this. (I thought we did already somewhere, but with a quick issue search I didn't find anything.)

File: mne/preprocessing/interpolate.py
Comment (larsoner) [inline mne/preprocessing/interpolate.py:111]: There are many int-like things, this will get you an `int` back which is nice
```suggestion
    bad_limit = _ensure_int(bad_limit, "bad_limit")
```

File: mne/preprocessing/interpolate.py
Comment (larsoner) [inline mne/preprocessing/interpolate.py:203]: Otherwise it seems like a standard centroid
```suggestion
def _find_centroid_sphere(ch_pos, group_names):
```

**Review by larsoner — COMMENTED** [review]
Just minor comments from me, otherwise LGTM!

**Comment by mscheltienne** [comment]
@larsoner I addressed your last comments, good to merge on my end.
Do you want to wait for another review, maybe @alexrockhill as he authored the original function?

**Comment by larsoner** [comment]
@mscheltienne I disabled the GitHub "notify on commits pushed" so I didn't see that you had addressed my comments (too many emails) so thanks for the ping! I think it's fine to merge as-is, but @alexrockhill let me know if you want to look

**Review by alexrockhill — APPROVED** [review]

**Pull request merged**
